pkg/recipes/terraform/config/config.go

```diff
@@ -38,18 +44,102 @@ const (
 
 // New creates TerraformConfig with the given module name and its inputs (module source, version, parameters)
 // Parameters are populated from environment recipe and resource recipe metadata.
-func New(moduleName string, envRecipe *recipes.EnvironmentDefinition, resourceRecipe *recipes.ResourceMetadata) *TerraformConfig {
+func New(ctx context.Context, moduleName string, envRecipe *recipes.EnvironmentDefinition, resourceRecipe *recipes.ResourceMetadata, envConfig *recipes.Configuration, armOptions *arm.ClientOptions) (*TerraformConfig, error) {
+	// Getting the module source path with credentials information if its a private source.
+	path, err := getModuleSource(ctx, envConfig, envRecipe.TemplatePath, armOptions)
+	if err != nil {
+		return nil, err
+	}
 	// Resource parameter gets precedence over environment level parameter,
 	// if same parameter is defined in both environment and resource recipe metadata.
-	moduleData := newModuleConfig(envRecipe.TemplatePath, envRecipe.TemplateVersion, envRecipe.Parameters, resourceRecipe.Parameters)
+	moduleData := newModuleConfig(path, envRecipe.TemplateVersion, envRecipe.Parameters, resourceRecipe.Parameters)
 
 	return &TerraformConfig{
 		Terraform: nil,
 		Provider:  nil,
 		Module: map[string]TFModuleConfig{
 			moduleName: moduleData,
 		},
+	}, nil
+}
+
+// getModuleSource is called to get the module source path with credential information
+// if template path represent a private source else returns the provided template path
+func getModuleSource(ctx context.Context, envConfig *recipes.Configuration, templatePath string, armOptions *arm.ClientOptions) (string, error) {
+	var source string
+	var err error
+	if strings.HasPrefix(templatePath, "git::") {
+		source, err = getGitSource(ctx, envConfig, templatePath, armOptions)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		source = templatePath
+	}
+	return source, err
+}
+
+// getGitSource creates a terraform module source in a generic git format with credential information
+// e.g: git::https://<username>:<pat>@<example.com>/<repo_path>
+func getGitSource(ctx context.Context, envConfig *recipes.Configuration, templatePath string, armOptions *arm.ClientOptions) (string, error) {
+	secretStore, err := getSecretStoreID(*envConfig, templatePath)
+	if err != nil {
+		return "", err
+	}
+	if secretStore == "" {
+		return templatePath, err
+	}
+
+	secretStoreID, err := resources.ParseResource(secretStore)
+	if err != nil {
+		return "", err
+	}
+	client, err := v20231001preview.NewSecretStoresClient(secretStoreID.RootScope(), &aztoken.AnonymousCredential{}, armOptions)
+	if err != nil {
+		return "", err
 	}
+	secrets, err := client.ListSecrets(ctx, secretStoreID.Name(), map[string]any{}, nil)
```

Comment:
I'm not sure if this is the right place for this. I'd lean towards keeping interactions with CoreRP bounded to driver, and everything within Terraform executor and downstream is responsible for interacting with Terraform. @rynowak do you have any strong preference?

Reply:
We discussed this in the design review, this will be moved to the config loader layer.